engine/consensus/approvals/core.go

```diff
@@ -0,0 +1,372 @@
+	if err != nil {
+		return fmt.Errorf("could not get block height for incorporated block %s: %w",
+			result.IncorporatedBlockID, err)
+	}
+
+	lastFinalizedBlockHeight := atomic.LoadUint64(&c.lastFinalizedBlockHeight)
+
+	// check if we are dealing with finalized block or an orphan
+	if incorporatedAtHeight <= lastFinalizedBlockHeight {
+		finalized, err := c.state.AtHeight(incorporatedAtHeight).Head()
+		if err != nil {
+			return fmt.Errorf("could not retrieve finalized block at height %d: %w", incorporatedAtHeight, err)
+		}
+		if finalized.ID() != result.IncorporatedBlockID {
+			// it means that we got incorporated result for a block which doesn't extend our chain
+			// and should be discarded from future processing
+			return engine.NewOutdatedInputErrorf("won't process incorporated result from orphan block %s", result.IncorporatedBlockID)
+		}
+	}
+
+	collector := c.getCollector(result.Result.ID())
+	newIncorporatedResult := collector == nil
+	if newIncorporatedResult {
+		// in case block is not finalized we will create collector and start processing approvals
+		// no checks for orphans can be made at this point
+		// we expect that assignment collector will cleanup orphan IRs whenever new finalized block is processed
+
+		collector, err = c.createCollector(result.Result)
+		if err != nil {
+			return fmt.Errorf("could not process incorporated result, cannot create collector: %w", err)
+		}
+	}
+
+	err = collector.ProcessIncorporatedResult(result)
+	if err != nil {
+		return fmt.Errorf("could not process incorporated result: %w", err)
+	}
+
+	// process pending approvals only if it's a new collector
+	if newIncorporatedResult {
+		err = c.processPendingApprovals(collector)
+		if err != nil {
+			return fmt.Errorf("could not process cached approvals:  %w", err)
+		}
+	}
+
+	return nil
+}
+
+// checkBlockOutdated performs a sanity check if block is outdated
+// Returns:
+// * engine.UnverifiableInputError - sentinel error in case we haven't discovered requested blockID
+// * engine.OutdatedInputError - sentinel error in case block is outdated
+// * exception in case of unknown internal error
+// * nil - block isn't sealed
+func (c *approvalProcessingCore) checkBlockOutdated(blockID flow.Identifier) error {
+	height, err := c.getCachedBlockHeight(blockID)
+	if err != nil {
+		return err
+	}
+
+	// it's important to use atomic operation to make sure that we have correct ordering
+	lastSealedHeight := atomic.LoadUint64(&c.lastSealedBlockHeight)
+	// drop approval, if it is for block whose height is lower or equal to already sealed height
+	if lastSealedHeight >= height {
+		return engine.NewOutdatedInputErrorf("result is for already sealed and finalized block height")
+	}
+
+	return nil
+}
+
+func (c *approvalProcessingCore) ProcessApproval(approval *flow.ResultApproval) error {
+	err := c.checkBlockOutdated(approval.Body.BlockID)
+	if err != nil {
+		return err
```

Comment:
better to wrap the error with approval id and block id, otherwise the error isn't useful for debugging